controllers/routemonitor/routemonitor_controller_supplement.go

```diff
@@ -17,175 +17,355 @@ import (
-	if routeMonitor.Status.RouteURL == "" {
-		return customerrors.NoHost
+func (r *RouteMonitorReconciler) CreateBlackBoxExporterResources(ctx context.Context) error {
+	if err := r.createBlackBoxExporterDeployment(ctx); err != nil {
+		return err
 	}
-	// serviceMonitorName is joined by the name and the namespace to create a unique ServiceMonitor for each RouteMonitor
-	serviceMonitorName := fmt.Sprintf("%s-%s", routeMonitor.Name, routeMonitor.Namespace)
-	namespacedName := types.NamespacedName{Name: serviceMonitorName, Namespace: blackBoxNamespace}
-	// PRE
+	// Creating Service after because:
+	//
+	// A Service should not point to an empty target (Deployment)
+	if err := r.createBlackBoxExporterService(ctx); err != nil {
+		return err
+	}
+	return nil
+}
 
-	serviceMonitor := &monitoringv1.ServiceMonitor{}
-	populationFunc := func() *monitoringv1.ServiceMonitor {
-		return r.templateForServiceMonitorResource(routeMonitor, serviceMonitorName)
+func (r *RouteMonitorReconciler) createBlackBoxExporterDeployment(ctx context.Context) error {
+	resource := &appsv1.Deployment{}
+	populationDeploymentFunc := func() appsv1.Deployment {
+		return r.templateForBlackBoxExporterDeployment()
 	}
 	// Does the resource already exist?
-	err := r.Get(ctx, namespacedName, serviceMonitor)
+	err := r.Get(ctx, blackBoxNamespacedName, resource)
 	if err != nil {
 		// If this is an unknown error
 		if !k8serrors.IsNotFound(err) {
 			// return unexpectedly
 			return err
 		}
 		// populate the resource with the template
-		resource := populationFunc()
+		resource := populationDeploymentFunc()
 		// and create it
-		err = r.Create(ctx, resource)
+		err = r.Create(ctx, &resource)
 		if err != nil {
 			return err
 		}
 	}
 	return nil
 }
 
-const (
-	blackBoxNamespace  = "openshift-monitoring"
-	blackBoxName       = "blackbox-exporter"
-	blackBoxPortName   = "blackbox"
-	blackBoxPortNumber = 9115
-)
-
-var ( // cannot be a const but doesn't ever change
-	blackBoxNamespacedName = types.NamespacedName{Name: blackBoxName, Namespace: blackBoxNamespace}
-)
-
-func (r *RouteMonitorReconciler) CreateBlackBoxExporterDeployment(ctx context.Context) error {
-	resource := &appsv1.Deployment{}
-	populationDeploymentFunc := func() *appsv1.Deployment {
-		return r.templateForBlackBoxExporterDeployment()
+func (r *RouteMonitorReconciler) createBlackBoxExporterService(ctx context.Context) error {
+	resource := &corev1.Service{}
+	populationServiceFunc := func() corev1.Service {
+		return r.templateForBlackBoxExporterService()
 	}
 	// Does the resource already exist?
-	err := r.Get(ctx, blackBoxNamespacedName, resource)
-	if err != nil {
+	if err := r.Get(ctx, blackBoxNamespacedName, resource); err != nil {
 		// If this is an unknown error
 		if !k8serrors.IsNotFound(err) {
 			// return unexpectedly
 			return err
 		}
 		// populate the resource with the template
-		resource := populationDeploymentFunc()
+		resource := populationServiceFunc()
```

Comment:
can't we call `r.templateForBlackBoxExporterService()` directly here?

Reply:
we can, but this makes the code more similar to other sections. this is also why I renamed the function name

so when someone reads this he can ignore the code that is similar and read only what changed